unified_planning/engines/compilers/grounder.py

```diff
@@ -32,6 +32,152 @@
+    it at construction time.
+
+    It offers the capability both of grounding a whole problem, like the :class:`~unified_planning.engines.compilers.Grounder`
+    in the :func:`~unified_planning.engines.compilers.Grounder.compile` method (with the :func:`~unified_planning.engines.compilers.GroundingSupport.ground_problem`
+    method) or offers the capability of grounding a single Action of the Problem given the grounding parameters
+    (with the :func:`~unified_planning.engines.compilers.GroundingSupport.ground_action` method).
+
+    This class also caches the grounded actions so it avoids duplication.
+    """
+
+    def __init__(
+        self,
+        problem: Problem,
+        grounding_actions_map: Optional[Dict[Action, List[Tuple[FNode, ...]]]] = None,
+    ):
+        assert isinstance(problem, Problem)
+        self._problem = problem
+        self._grounding_actions_map = grounding_actions_map
+        # grounded_actions is a map from an Action of the original problem and it's parameters
+        # to the grounded instance of the Action with the given parameters.
+        self._grounded_actions: Dict[
+            Tuple[Action, Tuple[FNode, ...]], Optional[Action]
+        ] = {}
+        env = problem.env
+        self._substituter = Substituter(env)
+        self._simplifier = Simplifier(env, problem)
+        self._grounding_result: Optional[CompilerResult] = None
+
+    @property
+    def name(self):
+        return "grounder"
+
+    def ground_action(
+        self, action: Action, parameters: Tuple[FNode, ...] = tuple()
+    ) -> Optional[Action]:
+        """
+        Grounds the given `action` with the given `parameters`.
+        An `Action` is grounded when it has no :func:`parameters <unified_planning.model.Action.parameters>`.
+
+        The returned `Action` is cached, so if the same method is called twice with the same function parameters,
+        the same object is returned, and the same object will be returned in the total problem grounding, so
+        if the resulting `Action` or :class:`~unified_planning.model.Problem` are modified, all the copies
+        returned by this class will be modified.
+
+        :param action: The `Action` that must be grounded with the given `parameters`.
+        :param parameters: The tuple of expressions used to ground the given `action`.
+        :return: The `action` grounded with the given `parameters` or `None` if the grounded
+            action does not have `effects` or the `action conditions` can be easily evaluated as a
+            contradiction.
+        """
+        assert len(action.parameters) == len(
+            parameters
+        ), "The number of given parameters for the grounding is different from the action's parameters"
+        key = (action, tuple(parameters))
+        if key in self._grounded_actions:  # The action is already created
+            return self._grounded_actions[key]
```

Comment:
I thought about this, and the way I would access the dictionary only once would be:

```
value = self._grounded_actions.get(key, None)
if value is not None:
     return value
```

The problem with this code is that `None` is actually a valid result for a grounded action. It means that the created grounded action does not have a "meaning" (it has no effects or the preconditions are a contradiction) (and here you also have the answer to the first comment "why Optional?" ) . So, given that `None` is a valid result, i couldn't use the code i wanted to use.. 

But now we have to discuss if `None` is actually a valid result for it's meaning or we want to change that; because as it is now, the method that creates the grounded action (`create_action_with_given_subs`) returns `None` if the action created is meaningless.